File: php/SchemaReader.php
<?php

namespace POO;

class SchemaReader
{
    protected $graph;

    public function __construct(string $path)
    {
        $jsonldSchema = file_get_contents($path);
        $data = json_decode($jsonldSchema, true);

        if ($data === null) {
            throw new Exception('Json decode error: 0x' . json_last_error() . ' - ' . json_last_error_msg());
        }

        if (!array_key_exists('@graph', $data)) {
            throw new Exception('Invalid input data. @graph not found.');
        }

        // indexing by @id
        $this->graph = array_combine(array_column($data['@graph'], '@id'), $data['@graph']);
    }

    function get(string $id)
    {
        return isset($this->graph[$id]) ? $this->graph[$id] : null;
    }

    function getSuperClasses(string $id)
    {
        $class = $this->get($id);
        
        if ($class === null || !array_key_exists('@type', $class) || $class['@type'] !== 'rdfs:Class') {
            return [];
        }

        $classes = [$id];

        // zero parents
        if (!array_key_exists('rdfs:subClassOf', $class)) {
            return $classes;
        }

        // one parent
        if (array_key_exists('@id', $class['rdfs:subClassOf'])) {
            $superSuper = $this->getSuperClasses($class['rdfs:subClassOf']['@id']);
            return array_merge($classes, $superSuper);
        }

        // multiple parents
        foreach ($class['rdfs:subClassOf'] as $superClass) {
            $superSuper = $this->getSuperClasses($superClass['@id']);
            $classes = array_merge($classes, $superSuper);
        }
        return array_unique($classes);
    }
}

File: python/SchemaReader.py
import json


class SchemaReader:
    def __init__(self, filepath):
        try:
            with open(filepath, 'r') as f:
                data = json.load(f)
        except ValueError:
            raise ValueError('Decoding JSON has failed')

        if '@graph' not in data:
            raise ValueError('Invalid input data. @graph not found.')

        self.graph = dict([(e['@id'], e) for e in data['@graph'] if '@id' in e])

    def get(self, id):
        return self.graph.get(id)

    def getSuperClasses(self, id):
        klass = self.graph[id] if id in self.graph else False

        if not klass or '@type' not in klass or klass['@type'] != 'rdfs:Class':
            return list()

        classes = list(id)

        # zero parents
        if 'rdfs:subClassOf' not in klass:
            return classes

        # one parent
        if '@id' in klass['rdfs:subClassOf']:
            superSuper = self.getSuperClasses(klass['rdfs:subClassOf']['@id'])
            return list(set(classes + superSuper))

        # multiple parents
        for key, superClass in klass['rdfs:subClassOf'].items():
            classes += self.getSuperClasses(superClass['@id'])

        return list(set(classes))

    def getTypes(self, id):
        return [e for e in self.graph.values() if e['@type'] == 'rdfs:Class']
        # return list(filter(lambda e: e['@type'] == 'rdfs:Class', self.graph.values()))

0
